src/JitInterface/src/SystemVStructClassificator.cs

```diff
@@ -0,0 +1,648 @@
+                }
+
+                if (fieldClassificationType == SystemVClassificationType.SystemVClassificationTypeStruct)
+                {
+                    bool inEmbeddedStructPrev = helper.inEmbeddedStruct;
+                    helper.inEmbeddedStruct = true;
+
+                    bool structRet = false;
+                    structRet = ClassifyEightBytes(field.FieldType, ref helper, normalizedFieldOffset);
+                    
+                    helper.inEmbeddedStruct = inEmbeddedStructPrev;
+
+                    if (!structRet)
+                    {
+                        // If the nested struct says not to enregister, there's no need to continue analyzing at this level. Just return do not enregister.
+                        return false;
+                    }
+
+                    continue;
+                }
+
+                if (fieldClassificationType == SystemVClassificationType.SystemVClassificationTypeTypedReference || 
+                    TypeDef2SystemVClassification(typeDesc) == SystemVClassificationType.SystemVClassificationTypeTypedReference)
+                {
+                    // The TypedReference is a very special type.
+                    // In source/metadata it has two fields - Type and Value and both are defined of type IntPtr.
+                    // When the VM creates a layout of the type it changes the type of the Value to ByRef type and the
+                    // type of the Type field is left to IntPtr (TYPE_I internally - native int type.)
+                    // This requires a special treatment of this type. The code below handles the both fields (and this entire type).
+
+                    for (int i = 0; i < 2; i++)
+                    {
+                        fieldSize = 8;
+                        fieldOffset = (i == 0 ? 0 : 8);
+                        normalizedFieldOffset = fieldOffset + startOffsetOfStruct;
+                        fieldClassificationType = (i == 0 ? SystemVClassificationType.SystemVClassificationTypeIntegerByRef : SystemVClassificationType.SystemVClassificationTypeInteger);
+                        if ((normalizedFieldOffset % fieldSize) != 0)
+                        {
+                            // The spec requires that struct values on the stack from register passed fields expects
+                            // those fields to be at their natural alignment.
+                            return false;
+                        }
+
+                        helper.largestFieldOffset = (int)normalizedFieldOffset;
+
+                        // Set the data for a new field.
+
+                        // The new field classification must not have been initialized yet.
+                        Debug.Assert(helper.fieldClassifications[helper.currentUniqueOffsetField] == SystemVClassificationType.SystemVClassificationTypeNoClass);
+
+                        // There are only a few field classifications that are allowed.
```

Comment:
15 lines above we set this variable to either Integer or IntegerByRef - there's little point in this assert.

Reply:
It seems you are right. Again, this is a copy of the coreclr stuff. I guess it might have been just left there after some refactoring of the code that happened few months ago.